File: python/src/fintech_tick_run_bars/streaming.py
"""Stateful, streaming Tick-Run-Bar builder.

:func:`~fintech_tick_run_bars.core.construct_bars` aggregates a whole tape at
once. A live tape needs a builder that accepts one trade at a time and emits a
bar the instant ``max(buyTicks, sellTicks) >= threshold``.

Because this bar type is *adaptive*, the builder also **exposes its state**:
:attr:`buy_ticks`, :attr:`sell_ticks`, :attr:`dominant_count`,
:attr:`threshold` (the frozen snapshot the open bar is judged against),
:attr:`expected_ticks`, :attr:`expected_buy_probability`, and :attr:`tick_sign`.

Feed trades with :meth:`push` (returns any bars closed by that trade — zero or
one); call :meth:`flush` at end of stream. Bars are byte-identical to the batch
kernel.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from .core import (
    REQUIRED_TRADE_FIELDS,
    TickRunBarsValidationError,
    _finite_number,
    _rounded,
    _timestamp,
    _validate_config,
)

__all__ = ["StreamingTickRunBarBuilder"]


class StreamingTickRunBarBuilder:
    """Incremental Tick-Run-Bar builder with observable adaptive state.

    Examples
    --------
    >>> builder = StreamingTickRunBarBuilder(config)        # doctest: +SKIP
    >>> closed = builder.push_many(tape) + builder.flush()  # doctest: +SKIP
    >>> builder.dominant_count, builder.threshold           # doctest: +SKIP
    """

    def __init__(self, config: dict[str, Any]) -> None:
        _validate_config(config)
        self._close_partial = bool(config["closePartial"])
        self._initial_sign = int(config["initialTickSign"])
        self._initial_expected_ticks = float(config["initialExpectedTicks"])
        self._initial_buy_probability = float(config["initialBuyProbability"])
        self._alpha_ticks = float(config["alphaTicks"])
        self._alpha_probability = float(config["alphaBuyProbability"])
        self._floor = float(config["thresholdFloorTicks"])
        self._multiplier = float(config["thresholdMultiplier"])

        # validation state
        self._ids: set[str] = set()
        self._closed_sessions: set[str] = set()
        self._validation_session: str | None = None
        self._prior_time: datetime | None = None
        self._prior_sequence: int | None = None
        self._symbol: str | None = None
        self._currency: str | None = None

        # adaptive + aggregation state
        self._current: list[dict[str, Any]] = []
        self._session: str | None = None
        self._previous_price: float | None = None
        self._tick_sign = self._initial_sign
        self._expected_ticks = self._initial_expected_ticks
        self._expected_buy_probability = self._initial_buy_probability
        self._buy_ticks = 0
        self._sell_ticks = 0
        self._threshold = 0.0
        self._frozen_expected_ticks = 0.0
        self._frozen_buy_probability = 0.0
        self._bar_count = 0
        self._flushed = False
        self._begin_bar()

    # -- observable state --------------------------------------------------- #
    @property
    def buy_ticks(self) -> int:
        """Cumulative ``+1`` signs in the open bar."""
        return self._buy_ticks

    @property
    def sell_ticks(self) -> int:
        """Cumulative ``-1`` signs in the open bar."""
        return self._sell_ticks

    @property
    def dominant_count(self) -> int:
        """The run length being tested: ``max(buy_ticks, sell_ticks)``."""
        return max(self._buy_ticks, self._sell_ticks)

    @property
    def threshold(self) -> float:
        """Frozen threshold (ticks) the open bar is being judged against."""
        return self._threshold

    @property
    def expected_ticks(self) -> float:
        """Live EWMA estimate of ticks per bar (feeds the *next* bar's threshold)."""
        return self._expected_ticks

    @property
    def expected_buy_probability(self) -> float:
        """Live EWMA estimate of the buy-tick probability."""
        return self._expected_buy_probability

    @property
    def tick_sign(self) -> int:
        """Sign the next flat trade would carry (+1 or -1)."""
        return self._tick_sign

    @property
    def tick_count(self) -> int:
        """Trades currently held in the open bar."""
        return len(self._current)

    # -- internals ---------------------------------------------------------- #
    def _begin_bar(self) -> None:
        self._current = []
        self._buy_ticks = 0
        self._sell_ticks = 0
        self._frozen_expected_ticks = self._expected_ticks
        self._frozen_buy_probability = self._expected_buy_probability
        self._threshold = max(
            self._floor,
            self._multiplier
            * self._frozen_expected_ticks
            * max(self._frozen_buy_probability, 1 - self._frozen_buy_probability),
        )

    def _validate_trade(self, trade: object) -> None:
        if not isinstance(trade, dict):
            raise TickRunBarsValidationError("each trade must be a dictionary")
        missing = [field for field in REQUIRED_TRADE_FIELDS if field not in trade]
        if missing:
            raise TickRunBarsValidationError(f"trade is missing: {', '.join(missing)}")
        for field in ("tradeId", "session", "symbol", "currency"):
            if not isinstance(trade[field], str) or not trade[field].strip():
                raise TickRunBarsValidationError(f"{field} must be a non-empty string")
        if trade["tradeId"] in self._ids:
            raise TickRunBarsValidationError("tradeId must be unique")
        self._ids.add(trade["tradeId"])

        current_time = _timestamp(trade["timestamp"])
        if self._prior_time is not None and current_time < self._prior_time:
            raise TickRunBarsValidationError("trades must be chronological")
        if self._prior_time is not None and current_time == self._prior_time:
            sequence = trade.get("sequence")
            if (
                isinstance(sequence, bool)
                or not isinstance(sequence, int)
                or self._prior_sequence is None
                or sequence <= self._prior_sequence
            ):
                raise TickRunBarsValidationError(
                    "equal timestamps require strictly increasing integer sequence values"
                )
        self._prior_sequence = (
            trade.get("sequence") if isinstance(trade.get("sequence"), int) else None
        )
        self._prior_time = current_time

        price = _finite_number(trade["price"], "price")
        volume = _finite_number(trade["volume"], "volume")
        if price <= 0 or volume <= 0:
            raise TickRunBarsValidationError("price and volume must be positive")

        if self._symbol is None:
            self._symbol, self._currency = trade["symbol"], trade["currency"]
        elif trade["symbol"] != self._symbol or trade["currency"] != self._currency:
            raise TickRunBarsValidationError("one symbol and one currency are allowed per call")

        if self._validation_session is None:
            self._validation_session = trade["session"]
        elif trade["session"] != self._validation_session:
            self._closed_sessions.add(self._validation_session)
            if trade["session"] in self._closed_sessions:
                raise TickRunBarsValidationError(
                    "a session may not reappear after another session begins"
                )
            self._validation_session = trade["session"]

    def _emit(self, reason: str) -> dict[str, Any] | None:
        if not self._current:
            return None
        prices = [float(t["price"]) for t in self._current]
        volumes = [float(t["volume"]) for t in self._current]
        dominant_count = max(self._buy_ticks, self._sell_ticks)
        dominant_side = (
            "buy"
            if self._buy_ticks > self._sell_ticks
            else "sell" if self._sell_ticks > self._buy_ticks else "tie"
        )
        is_complete = reason == "threshold"
        bar = {
            "barIndex": self._bar_count,
            "session": self._current[0]["session"],
            "startTime": self._current[0]["timestamp"],
            "endTime": self._current[-1]["timestamp"],
            "lastTradeTime": self._current[-1]["timestamp"],
            "open": _rounded(prices[0]),
            "high": _rounded(max(prices)),
            "low": _rounded(min(prices)),
            "close": _rounded(prices[-1]),
            "volume": _rounded(sum(volumes)),
            "dollarValue": _rounded(sum(p * v for p, v in zip(prices, volumes))),
            "tickCount": len(self._current),
            "firstTradeId": self._current[0]["tradeId"],
            "lastTradeId": self._current[-1]["tradeId"],
            "closeReason": reason,
            "buyTicks": self._buy_ticks,
            "sellTicks": self._sell_ticks,
            "dominantCount": dominant_count,
            "dominantSide": dominant_side,
            "thresholdTicks": _rounded(self._threshold),
            "overshootTicks": _rounded(max(0.0, dominant_count - self._threshold))
            if is_complete
            else 0.0,
            "thresholdMet": is_complete,
            "isComplete": is_complete,
            "frozenExpectedTicks": _rounded(self._frozen_expected_ticks),
            "frozenBuyProbability": _rounded(self._frozen_buy_probability),
        }
        self._bar_count += 1
        if is_complete:
            observed_ticks = len(self._current)
            observed_buy_probability = self._buy_ticks / observed_ticks
            self._expected_ticks = (
                1 - self._alpha_ticks
            ) * self._expected_ticks + self._alpha_ticks * observed_ticks
            self._expected_buy_probability = (
                1 - self._alpha_probability
            ) * self._expected_buy_probability + self._alpha_probability * observed_buy_probability
        self._begin_bar()
        return bar

    # -- public API --------------------------------------------------------- #
    def push(self, trade: dict[str, Any]) -> list[dict[str, Any]]:
        """Accept one trade and return any bars it closes (zero or one)."""
        if self._flushed:
            raise TickRunBarsValidationError("cannot push after flush()")
        self._validate_trade(trade)
        emitted: list[dict[str, Any]] = []

        if self._session is not None and trade["session"] != self._session:
            if self._current and self._close_partial:
                bar = self._emit("session_end")
                if bar is not None:
                    emitted.append(bar)
            else:
                self._begin_bar()
            # Session reset: price, sign, and both expectations return to seeds.
            self._previous_price = None
            self._tick_sign = self._initial_sign
            self._expected_ticks = self._initial_expected_ticks
            self._expected_buy_probability = self._initial_buy_probability
            self._begin_bar()
        self._session = trade["session"]

        price = float(trade["price"])
        if self._previous_price is not None:
            if price > self._previous_price:
                self._tick_sign = 1
            elif price < self._previous_price:
                self._tick_sign = -1
        self._previous_price = price
        self._current.append(trade)
        self._buy_ticks += int(self._tick_sign == 1)
        self._sell_ticks += int(self._tick_sign == -1)

        if max(self._buy_ticks, self._sell_ticks) >= self._threshold:
            bar = self._emit("threshold")
            if bar is not None:
                emitted.append(bar)
        return emitted

    def push_many(self, trades: object) -> list[dict[str, Any]]:
        """Feed an iterable of trades, returning every bar closed along the way."""
        try:
            iterator = iter(trades)  # type: ignore[arg-type]
        except TypeError as error:
            raise TickRunBarsValidationError("trades must be an iterable.") from error
        emitted: list[dict[str, Any]] = []
        for trade in iterator:
            emitted.extend(self.push(trade))
        return emitted

    def flush(self) -> list[dict[str, Any]]:
        """Close the final partial bar (if ``closePartial``) and end the stream."""
        self._flushed = True
        if self._current and self._close_partial:
            bar = self._emit("stream_end")
            return [bar] if bar is not None else []
        return []

File: typescript/src/streaming.ts
/**
 * Stateful, streaming Tick-Run-Bar builder.
 *
 * `./bars`'s `constructBars` aggregates a whole tape at once. A live tape needs a
 * builder that accepts one trade at a time and emits a bar the instant
 * `max(buyTicks, sellTicks) >= threshold`.
 *
 * Because this bar type is *adaptive*, the builder also **exposes its state**:
 * `buyTicks`, `sellTicks`, `dominantCount`, `threshold` (the frozen snapshot the
 * open bar is judged against), `expectedTicks`, `expectedBuyProbability`, and
 * `tickSign`.
 */

import {
  REQUIRED_TRADE_FIELDS,
  TickRunBarsValidationError,
  finite,
  rounded,
  timestampMs,
  validateConfig,
  type Bar,
  type CloseReason,
  type Config,
  type DominantSide,
  type Trade,
} from "./bars.ts";

export class StreamingTickRunBarBuilder {
  private readonly closePartial: boolean;
  private readonly initialSign: number;
  private readonly initialExpectedTicks: number;
  private readonly initialBuyProbability: number;
  private readonly alphaTicks: number;
  private readonly alphaProbability: number;
  private readonly floor: number;
  private readonly multiplier: number;

  // validation state
  private ids = new Set<string>();
  private closedSessions = new Set<string>();
  private validationSession: string | null = null;
  private priorTime: number | null = null;
  private priorSequence: number | null = null;
  private symbol: string | null = null;
  private currency: string | null = null;

  // adaptive + aggregation state
  private current: Trade[] = [];
  private session: string | null = null;
  private previousPrice: number | null = null;
  private _tickSign: number;
  private _expectedTicks: number;
  private _expectedBuyProbability: number;
  private _buyTicks = 0;
  private _sellTicks = 0;
  private _threshold = 0;
  private frozenExpectedTicks = 0;
  private frozenBuyProbability = 0;
  private barCount = 0;
  private flushed = false;

  constructor(config: Config) {
    validateConfig(config);
    this.closePartial = config.closePartial;
    this.initialSign = config.initialTickSign;
    this.initialExpectedTicks = config.initialExpectedTicks;
    this.initialBuyProbability = config.initialBuyProbability;
    this.alphaTicks = config.alphaTicks;
    this.alphaProbability = config.alphaBuyProbability;
    this.floor = config.thresholdFloorTicks;
    this.multiplier = config.thresholdMultiplier;
    this._tickSign = this.initialSign;
    this._expectedTicks = this.initialExpectedTicks;
    this._expectedBuyProbability = this.initialBuyProbability;
    this.beginBar();
  }

  /** Cumulative `+1` signs in the open bar. */
  get buyTicks(): number {
    return this._buyTicks;
  }

  /** Cumulative `-1` signs in the open bar. */
  get sellTicks(): number {
    return this._sellTicks;
  }

  /** The run length being tested: `max(buyTicks, sellTicks)`. */
  get dominantCount(): number {
    return Math.max(this._buyTicks, this._sellTicks);
  }

  /** Frozen threshold (ticks) the open bar is being judged against. */
  get threshold(): number {
    return this._threshold;
  }

  /** Live EWMA estimate of ticks per bar (feeds the *next* bar's threshold). */
  get expectedTicks(): number {
    return this._expectedTicks;
  }

  /** Live EWMA estimate of the buy-tick probability. */
  get expectedBuyProbability(): number {
    return this._expectedBuyProbability;
  }

  /** Sign the next flat trade would carry (+1 or -1). */
  get tickSign(): number {
    return this._tickSign;
  }

  /** Trades currently held in the open bar. */
  get tickCount(): number {
    return this.current.length;
  }

  private beginBar(): void {
    this.current = [];
    this._buyTicks = 0;
    this._sellTicks = 0;
    this.frozenExpectedTicks = this._expectedTicks;
    this.frozenBuyProbability = this._expectedBuyProbability;
    this._threshold = Math.max(
      this.floor,
      this.multiplier *
        this.frozenExpectedTicks *
        Math.max(this.frozenBuyProbability, 1 - this.frozenBuyProbability),
    );
  }

  private validateTrade(raw: unknown): Trade {
    if (!raw || typeof raw !== "object") {
      throw new TickRunBarsValidationError("each trade must be an object");
    }
    const trade = raw as Trade;
    const missing = REQUIRED_TRADE_FIELDS.filter((field) => !(field in trade));
    if (missing.length) {
      throw new TickRunBarsValidationError(`trade is missing: ${missing.join(", ")}`);
    }
    for (const field of ["tradeId", "session", "symbol", "currency"] as const) {
      if (typeof trade[field] !== "string" || !trade[field].trim()) {
        throw new TickRunBarsValidationError(`${field} must be a non-empty string`);
      }
    }
    if (this.ids.has(trade.tradeId)) throw new TickRunBarsValidationError("tradeId must be unique");
    this.ids.add(trade.tradeId);

    const currentTime = timestampMs(trade.timestamp);
    if (this.priorTime !== null && currentTime < this.priorTime) {
      throw new TickRunBarsValidationError("trades must be chronological");
    }
    if (this.priorTime !== null && currentTime === this.priorTime) {
      const sequence = trade.sequence;
      if (
        !Number.isInteger(sequence) ||
        this.priorSequence === null ||
        (sequence as number) <= this.priorSequence
      ) {
        throw new TickRunBarsValidationError(
          "equal timestamps require strictly increasing integer sequence values",
        );
      }
    }
    this.priorSequence = Number.isInteger(trade.sequence) ? (trade.sequence as number) : null;
    this.priorTime = currentTime;

    const price = finite(trade.price, "price");
    const volume = finite(trade.volume, "volume");
    if (price <= 0 || volume <= 0) {
      throw new TickRunBarsValidationError("price and volume must be positive");
    }

    if (this.symbol === null) {
      this.symbol = trade.symbol;
      this.currency = trade.currency;
    } else if (trade.symbol !== this.symbol || trade.currency !== this.currency) {
      throw new TickRunBarsValidationError("one symbol and one currency are allowed per call");
    }

    if (this.validationSession === null) this.validationSession = trade.session;
    else if (trade.session !== this.validationSession) {
      this.closedSessions.add(this.validationSession);
      if (this.closedSessions.has(trade.session)) {
        throw new TickRunBarsValidationError("a session may not reappear after another session begins");
      }
      this.validationSession = trade.session;
    }
    return trade;
  }

  private emit(reason: CloseReason): Bar | null {
    if (!this.current.length) return null;
    const prices = this.current.map((t) => t.price);
    const volumes = this.current.map((t) => t.volume);
    const dominantCount = Math.max(this._buyTicks, this._sellTicks);
    const dominantSide: DominantSide =
      this._buyTicks > this._sellTicks ? "buy" : this._sellTicks > this._buyTicks ? "sell" : "tie";
    const isComplete = reason === "threshold";
    const bar: Bar = {
      barIndex: this.barCount,
      session: this.current[0].session,
      startTime: this.current[0].timestamp,
      endTime: this.current.at(-1)!.timestamp,
      lastTradeTime: this.current.at(-1)!.timestamp,
      open: rounded(prices[0]),
      high: rounded(Math.max(...prices)),
      low: rounded(Math.min(...prices)),
      close: rounded(prices.at(-1)!),
      volume: rounded(volumes.reduce((a, b) => a + b, 0)),
      dollarValue: rounded(this.current.reduce((a, t) => a + t.price * t.volume, 0)),
      tickCount: this.current.length,
      firstTradeId: this.current[0].tradeId,
      lastTradeId: this.current.at(-1)!.tradeId,
      closeReason: reason,
      buyTicks: this._buyTicks,
      sellTicks: this._sellTicks,
      dominantCount,
      dominantSide,
      thresholdTicks: rounded(this._threshold),
      overshootTicks: isComplete ? rounded(Math.max(0, dominantCount - this._threshold)) : 0,
      thresholdMet: isComplete,
      isComplete,
      frozenExpectedTicks: rounded(this.frozenExpectedTicks),
      frozenBuyProbability: rounded(this.frozenBuyProbability),
    };
    this.barCount += 1;
    if (isComplete) {
      const observedTicks = this.current.length;
      const observedBuyProbability = this._buyTicks / observedTicks;
      this._expectedTicks = (1 - this.alphaTicks) * this._expectedTicks + this.alphaTicks * observedTicks;
      this._expectedBuyProbability =
        (1 - this.alphaProbability) * this._expectedBuyProbability +
        this.alphaProbability * observedBuyProbability;
    }
    this.beginBar();
    return bar;
  }

  /** Accept one trade and return any bars it closes (zero or one). */
  push(rawTrade: Trade): Bar[] {
    if (this.flushed) throw new TickRunBarsValidationError("cannot push after flush()");
    const trade = this.validateTrade(rawTrade);
    const emitted: Bar[] = [];

    if (this.session !== null && trade.session !== this.session) {
      if (this.current.length && this.closePartial) {
        const bar = this.emit("session_end");
        if (bar) emitted.push(bar);
      } else {
        this.beginBar();
      }
      this.previousPrice = null;
      this._tickSign = this.initialSign;
      this._expectedTicks = this.initialExpectedTicks;
      this._expectedBuyProbability = this.initialBuyProbability;
      this.beginBar();
    }
    this.session = trade.session;

    if (this.previousPrice !== null) {
      this._tickSign =
        trade.price > this.previousPrice ? 1 : trade.price < this.previousPrice ? -1 : this._tickSign;
    }
    this.previousPrice = trade.price;
    this.current.push(trade);
    if (this._tickSign === 1) this._buyTicks += 1;
    if (this._tickSign === -1) this._sellTicks += 1;

    if (Math.max(this._buyTicks, this._sellTicks) >= this._threshold) {
      const bar = this.emit("threshold");
      if (bar) emitted.push(bar);
    }
    return emitted;
  }

  /** Feed an array of trades, returning every bar closed along the way. */
  pushMany(trades: readonly Trade[]): Bar[] {
    const emitted: Bar[] = [];
    for (const trade of trades) emitted.push(...this.push(trade));
    return emitted;
  }

  /** Close the final partial bar (if `closePartial`) and end the stream. */
  flush(): Bar[] {
    this.flushed = true;
    if (this.current.length && this.closePartial) {
      const bar = this.emit("stream_end");
      return bar ? [bar] : [];
    }
    return [];
  }
}
